### PR TITLE
SQL: SYS COLUMNS returns ODBC specific schema

### LINKED ISSUE #35376: ODBC: SYS COLUMNS column types alignment
`SYS COLUMNS` maps on the ODBC catalog function [`SQLColumns()`](https://docs.microsoft.com/en-us/sql/odbc/reference/syntax/sqlcolumns-function). This function prepares a table for the application to subsequently fetch. The table definition is given within the spec.

Current `SYS COLUMNS` implementation is aligned with the spec with the exception of the following columns, that are all returned as `integer`, while the application expects them as "Smallint" (i.e. SQL `SHORT`): `DATA_TYPE`, `DECIMAL_DIGITS `, `NUM_PREC_RADIX`, `NULLABLE`, `SQL_DATA_TYPE`, `SQL_DATETIME_SUB`.

This misalignment is problematic for the applications that:
- don't specify a particular type for the driver to convert the result to, but leave it to "default" (meaning "whatever type is in the DB"), but who also
- provide a buffer for the result which is aligned to the standard.

For example: when fetching `DATA_TYPE` value, the application provides a two byte buffer (to accommodate the `SHORT`). Leaving the conversion to "default", the driver would then try to access a four byte buffer, corresponding to DB's `INTEGER`, which eventually leads to a failure.

Now, this application behavior is not really correct (and [the spec advises against](https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/default-c-data-types) using "default" data type, but this is too late for 2.x applications). However, this problem can be mitigated if we aligned our types with the spec.

### PR DESCRIPTION
**Pull request by costin**
Due to some unresolvable type conflict between the expected definition
 in JDBC vs ODBC, the driver mode is now passed over so that certain
 command can change their results accordingly (in this case SYS COLUMNS)

Fix #35376

### CODE REVIEW

**Comment by costin** [comment]
run the gradle build tests 1

**Review by matriv — APPROVED** [review]
LGTM, left cosmetic comment.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysColumnsTests.java [hunk 1]
```diff
@@ -77,4 +141,20 @@ private static Object bufferLength(List<?> list) {
     private static Object radix(List<?> list) {
         return list.get(9);
     }
-}
+
+    private static Object nullable(List<?> list) {
+        return list.get(10);
+    }
+
+    private static Object decimalPrecision(List<?> list) {
```
[the remark this answers is not in the document]
Comment (matriv) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysColumnsTests.java]: It's really minor, I mean `decimalPrecision` could be before `radix` which is list.get(8).

**Comment by matriv** [comment]
The issue you're referring to, is not correct.

**Review by astefan — APPROVED** [review]
LGTM

**Pull request merged**